### ISSUE TITLE
Pagination for contacts within the selected contact tab

### ISSUE THREAD

**Issue by kennsippell** [medic/cht-core#8627]
**Describe the performance issue**
In this [forum post](https://forum.communityhealthtoolkit.org/t/cht-app-slowing-down/3041/10), performance experiments took ~57seconds to load a place's contact summary but then took another ~55seconds to render the contact list.

**Describe the improvement you'd like**
Consider pagination of the contacts list in the selected contacts tab

**To Reproduce**
Steps to record the performance metrics:
1. Create a place with 5500 contacts under it
1. Load the contact

**Measurements**
Contacts load at 0:07s
Contact summary loads at 0:57s
When I scroll down, I just see white. The list of contacts actually renders at 1:51s

Hamburger is very slow to open
Tab takes ~4 seconds to unload

**Environment**
- Instance: https://itechcares-malawi.dev.medicmobile.org/
- Browser: Chome117
- Client platform: Ubuntu
- App: webapp
- Version: 4.4.1 prerelease

**Comment by dianabarsan** [comment]
I believe we used to have a setting to prevent display of child contacts based on contact type, which _could_ be an option here. 
I think pagination is the best approach, but disabling display of contacts via config can be a short-term solution if this is blocking a project. 
Are you expecting all contacts of this type to have so many child contacts?

**Comment by kennsippell** [comment]
I had a hard time reproducing the ~2 minute load times yesterday; I'm mostly seeing ~25second load times now for the same scenario. I saw 2mins at least 4x in a row (not sure what's changed). One thing that did change is that performance improved in an incognito browser - my local extensions may have been part of the 2min load times. Sorry about that.  

I made a very quick PR #8629 to render only the first 50 contacts within any contact-content pane. My goal was just to quickly see what impact pagination might have for the CHT CARES app.  The timing in `contacts-content.component.ts` from `ngOnInit` to firing of the subscribe event in `subscriptionSelectedContactForms` went down from 10-14 seconds to 2-3 seconds. Note this is only the impact of reduced rendering (pagination would also reduce the cost of db interactions).

> I believe we used to have a setting to prevent display of child contacts based on contact type, which could be an option here.

This must have been removed, right? I don't see any permissions or setting related to this. Would be useful!

> Are you expecting all contacts of this type to have so many child contacts?

There is just one like this. It is a worst-case scenario. Disabling display of contacts would work for this case. This is the top-level facility, so really there is no reason to ever select this contact. But it is selected by default on devices with larger screens.

**Comment by dianabarsan** [comment]
> This must have been removed, right? I don't see any permissions or setting related to this. Would be useful!

Not _exactly_ removed. 
We [don't load child places if the opened contact is the logged in user's facility](https://github.com/medic/cht-core/blob/9cd6975e9d7f245e2d3c38cb307bb078a0b8d7f6/webapp/src/ts/effects/contacts.effects.ts#L119C28-L119C42). This is further down used here: https://github.com/medic/cht-core/blob/9cd6975e9d7f245e2d3c38cb307bb078a0b8d7f6/webapp/src/ts/services/contact-view-model-generator.service.ts#L199 

We kinda talked about this before: https://github.com/medic/cht-core/issues/4445

**Comment by latin-panda** [comment]
@michaelkohn and I had a quick look yesterday, and we're getting more info about this project, user behavior, etc... It likely involves some UI/UX design. We'll post back here with news.

**Comment by garethbowen** [comment]
@latin-panda @michaelkohn As this is currently impacting projects a compromise would be to do a quick and dirty pagination for 4.5.0 and then take time to do the proper UI/UX research and improve the solution for 4.6.0. If we're careful about setting the pagination size then it's likely that 99% of users will never see the implementation, and the 1% will be grateful the page loads at all.

**Comment by michaelkohn** [comment]
@garethbowen is there another project that you know of that has this problem?

Based on our discussion with the Cares Team last week, my understanding is a bit different than what is articulated above. We were told / shown that...

- The only users are tablet users, they normally work in portrait mode and normally only see single pane view
- They actually only have one place contact type (health center), the rest are person contact types
- The number of health centers is currently (i think) about 2 or 3
- There is a [meaningful way](https://forum.communityhealthtoolkit.org/t/cht-app-slowing-down/3041/16?u=michael) to split the patient contact types into additional places so that this issue disappears

Having said that... I think there might be a reason for Nurses to select that top level place (filling the Drug Collection and Facility Summary forms), but we didn't get a chance to talk about it so I'm not 100% sure.

Also FWIW, CARES will be upgrading to 4.4.x and then to a feature release to get the barcode stuff that @latin-panda is working on, so if we still need to do some dirty pagination, could we just include it in the feature release (and not 4.5)?

**Comment by garethbowen** [comment]
@michaelkohn Thanks the update!

> is there another project that you know of that has this problem?

None that I know of but with more projects scaling in ways we didn't design for I wouldn't be surprised if this is happening somewhere or will do soon.

@kennsippell Have you come across this issue anywhere else?

> The only users are tablet users, they normally work in portrait mode and normally only see single pane view

I think it's the single pane view that's the issue. My understanding is loading the RHS is slow regardless of whether you have 1 column or two.

> There is a [meaningful way](https://forum.communityhealthtoolkit.org/t/cht-app-slowing-down/3041/16?u=michael) to split the patient contact types into additional places so that this issue disappears

Changing a hierarchy is difficult, so unless it gives other benefits, improving rendering or pagination is a better solution.

> could we just include it in the feature release (and not 4.5)?

For bugs the point is to fix them in trunk so nobody else hits them either. It could also get backported to a feature release if desired. This still feels high priority to me which suggests we should prioritise fixing it in 4.5.0. What do you think?

**Comment by michaelkohn** [comment]
> I think it's the single pane view that's the issue. My understanding is loading the RHS is slow regardless of whether you have 1 column or two.

Yeah totally agree, RHS when tapped on that top level place is always going to be slow with their configuration, but single pane is fine because it doesn't need to render the RHS, and Nurses normally never need to see the RHS of that top level place so it is not clear to me how often anyone would experience this (hence Kenn's comment _"This is the top-level facility, so really there is no reason to ever select this contact. But it is selected by default on devices with larger screens."_

> Changing a hierarchy is difficult, so unless it gives other benefits, improving rendering or pagination is a better solution.

Isn't it going to be the exact same problem for loading reports and tasks too? I know we have some filters on those sections (1 & 2 week for Tasks, 3 and 6 months for Reports)... but 3 months worth of reports for 5,500 contacts is still a lot... probably 2 or 3 x5,500 (I know they do - or plan to do - some purging).

> For bugs the point is to fix them in trunk so nobody else hits them either 

Totally. I'd like to get @n-orlowski 's input first... I'm just not totally sold on the pagination solution just for Contacts on the RHS as the right "quick and dirty" solution. These users already see their contacts on the LHS, and they are paginated. To me a quicker (and dirtier? 😄 ) solution might be to completely remove contacts on the RHS if they are already visible on the LHS  - just like we did for places (though my question/concern about Tasks and Reports remains)... or have a setting on the contact_type that says whether or not to show those sections.

**Comment by garethbowen** [comment]
> Isn't it going to be the exact same problem for loading reports and tasks too?

Yeah quite likely. We need to design the whole page for scale.

> I'm just not totally sold on the pagination solution

Agreed, I don't think we should just jump to pagination as the solution as there might be something that works better. The nice thing about pagination is it's a catch all for if someone does something crazy with the configuration. If you had pagination set to 100 (or even just truncation with a "cannot show any more" statement) then when we do implement the better solution they work together. Say if we did "have a setting on the contact_type" but this was misconfigured for some users, then the pagination would save them from not being able to load the content.

We know that configuration is hard to get right, and we often don't hear details from users when something painful is happening ("the app is just slow!"), so it would make a safety net for when things go wrong.

**Comment by garethbowen** [comment]
It seems like this isn't blocking any deployment that we know of, and the workaround is fine for now. Bumping it to 4.6.0 to give us time to solve this properly.

**Comment by garethbowen** [comment]
@michaelkohn Should this block 4.6.0, move to 4.7.0, or go to the backlog?

**Comment by michaelkohn** [comment]
It should not block 4.6, we can move it to the backlog as there is adjacent work going on related to performance.
